### ISSUE TITLE
getSession() hangs indefinitely every hour or so

### ISSUE THREAD

**Issue by izhan** [supabase/gotrue-js#762]
# Bug report

- [X] I confirm this is a bug with Supabase, not with my own application. 
- [X] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

`await supabase.auth.getSession()` will hang indefinitely every hour or so, until the user reloads the page. We suspect it's related to the token needing to be refreshed.

## To Reproduce

1. Sign into an application with Supabase auth running
2. Wait an hour (or however long it takes for your token to expire)
3. Reload the page and call `await supabase.auth.getSession()`
4. `getSession` will hang indefinitely

It's unclear right now whether (4) happens all the time or only some of the time. But we've been managing to reproduce pretty frequently, both locally and in prod. It's been affecting our customers as well.

Reloading the browser after (4) will immediately resolve the issue. Subsequent reloads will be fine, until an hour of inactivity in which this problem occurs again.

We've tried multiple versions of supabase-js and gotrue already, to no effect. This has been occurring for the past few weeks.

https://github.com/supabase/supabase/issues/15930 seems related but that ticket seems to be concerned with immediately after Google OAuth. Our problem occurs well after initial login.

## Expected behavior

We expect `getSession` to never hang indefinitely

## System information

- OS: [e.g. macOS, Windows] macOS
- Browser (if applies) [e.g. chrome, safari] Chrome
- Version of supabase-js: [e.g. 6.0.2] Multiple (2.24, 2.26, 2.31)
- Version of Node.js: [e.g. 10.10.0] v18.16.0

## Additional context

We tried this on multiple versions of supabase-js (2.24, 2.26, 2.31) and gotrue (2.42, 2.43.1, 2.46.1)

**Comment by izhan** [comment]
@aviadmini the downgrade did work for me. Similar hang here too with the newer version, but I don't see any failed network requests, do you? It seemed to be a bug in the gotrue logic, ie when to actually kickoff the call to backend

**Comment by aviadmini** [comment]
When supabase is unreachable (e.g. local dev setup is down) there are console errors for the unsuccessful fetch and retries during a call to svelte auth helpers functions. Right after that the following call to `await supabase.auth.getSession()` never completes

**Comment by danrasmuson** [comment]
We experienced this issue...

<img width="1946" alt="CleanShot 2023-09-04 at 13 21 34@2x" src="https://github.com/supabase/gotrue-js/assets/4013767/14ccd4f2-ae6c-43b6-b097-87652fd13345">


Our call `supabaseClientSide.auth.getSession().then(({ data: { session } }) => session);` would just hang forever under certain circumstances when the client did not have the token.

The way our team was able to resolve this was by upgrading from 

    "@supabase/supabase-js": "^2.26.0"

to...

    "@supabase/supabase-js": "^2.33.1"

**Comment by andreasfelix** [comment]
We are also experiencing as similar issue when updating from supabase-js `2.30` -> `2.31`. the only change between these versions seems to be https://github.com/supabase/supabase-js/commit/c07ce232bf178c5abc9e5c3b096764b2807407b4 which updates `"@supabase/gotrue-js` from `2.43.1` to `2.46.1`

**Comment by kangmingtay** [comment]
hey everyone, i'm trying to reproduce this so that i can figure out why it happens but i need some help here - 

@andreasfelix you mentioned that:
> when awaiting a supabase function within the onAuthStateChange. The callback hangs indefinitely, depending on how fast the Promises returns. For us it only happens if the supabase-js client points to a local SUPABASE_URL setup, but not with a remote SUPABASE_URL...

which function are you calling in `onAuthStateChange` that leads to this?

@kjhughes just wondering, why do you need to call `getSession()` inside of the callback for `onAuthStateChange()`? The `onAuthStateChange()` [method](https://github.com/supabase/gotrue-js/blob/58a1ee69b566987052596e269ee8a3eef477dc2e/src/GoTrueClient.ts#L1458) will pass the event and the session as parameters to the callback function so there shouldn't be a need to call `getSession()` again

**Comment by andreasfelix** [comment]
> which function are you calling in `onAuthStateChange` that leads to this?

if I recall correctly any async operation that involves the supabase client, e.g.

```js
supabase.auth.onAuthStateChange(async (event, session) => {
  switch (event) {
    case "TOKEN_REFRESHED":
    case "SIGNED_IN":
        await supabase.rpc(...)
    ...
}
```

Our current workaround is to set a reactive variable (svelte) and run the effect whenever it changes:

```js
supabase.auth.onAuthStateChange((event, session) => {
  switch (event) {
    case "TOKEN_REFRESHED":
    case "SIGNED_IN":
      dirty.set(true)
    ...
}

...

dirty.subscribe(async ($dirty) => {
  if ($dirty) {
    dirty.set(false)
    await supabase.rpc(...)
  }
})
```

We didn't bother to look into why this works. I assume this somehow changes the order of microtasks changes (I am not super familiar with JS...).

**Comment by kangmingtay** [comment]
**TLDR: We are looking into a fix for this but for now the workaround is to **not** make async calls to postgREST / storage / edge functions using the supabase client lib inside the callback function registered in `onAuthStateChange` :/ **

Hey everyone, it seems like this issue was introduced when [this PR](https://github.com/supabase/gotrue-js/pull/757) was merged. For context, this PR was made to address issues related to random logouts where `_useSession` is being called concurrently which led to refresh tokens being unintentionally reused. The solution was to introduce a locking mechanism to serialise the calls made to `_useSession`. 

When an async supabase call is made in the callback function in `onAuthStateChange`, it results in a deadlock. This is because the supabase-js client library creates a [custom fetch method](https://github.com/supabase/supabase-js/blob/90b43533b754007c9a6fa5b71346d73bc4d9e410/src/lib/fetch.ts#L26) which makes a call to `_useSession` under the hood. 

This is how the deadlock occurs. For example, if you set up a callback function in `onAuthStateChange` to retrieve the latest todos on a `TOKEN_REFRESHED` event:
```
supabase.auth.onAuthStateChange( async (event, session) => {
  if (event === 'TOKEN_REFRESHED') {
    const todos = await supabase.from('todos').select('*')
  }
})
```
<img width="974" alt="image" src="https://github.com/supabase/gotrue-js/assets/28647601/e8a24365-fcec-4dc2-b092-8c255a0b44a0">
<img width="827" alt="image" src="https://github.com/supabase/gotrue-js/assets/28647601/e6649d59-a0b8-4641-af47-b13577512337">

If you make any async calls to postgREST / storage / functions using the supabase client library, it uses the custom fetch method mentioned above which requires a lock to be available.
![image](https://github.com/supabase/gotrue-js/assets/28647601/589832d7-468d-4903-a47d-356ebe0b79d5)

**Comment by izhan** [comment]
Thank you @kangmingtay for the update and @kjhughes for the helpful hints!

Confirmed on our end too that the bug we've been experiencing was caused by this. We shipped a workaround last week to avoid the async call on TOKEN_REFRESH and no longer see these issues in prod:
<img width="688" alt="CleanShot 2023-10-25 at 14 52 18@2x" src="https://github.com/supabase/gotrue-js/assets/4621962/bc730fc4-7baa-43ac-9872-287f383a765f">

**Comment by ryonwhyte** [comment]
Hi, I had the issue in React Native with @supabase/supabase-js": "^2.38.4"
Here is what the code that triggered the bug looked like 

```
 supabase.auth.onAuthStateChange(async (event, session) => {
      console.log(`Supabase auth event is ${event}`);
      //dispatch(setUserSession(session));
      //dispatch(setUser(session?.user ?? null));

      /**After an authStateChange we fetch user details. Maybe this should be conditional */
      if (session?.user?.id) {
        await new Promise((resolve) => setTimeout(resolve, 5000)); // Introduce a 1-second delay

        const details = await getProfileDetails(session.user.id);
        console.log("User details:", details);

        if (details.status === true) {
          dispatch(setProfile(details.data));
        }

        //Here we get the users orders
        const allUserOrders = await getUserOrders(session.user.id);
        //console.log("All user orders is: ", allUserOrders.data);
        if (allUserOrders.status === true) {
          dispatch(setUserOrders(allUserOrders.data));
        }
      }
    });
```

After getting some advice from @GaryAustin1 on discord. I removed all the functions inside onAuthStateChange functions and put them in another function, all made a single call. 

The code looked like this now and works fine. Note, all the logic was moved to external function "handleUserUpdates()"
```
supabase.auth.onAuthStateChange(async (event, session) => {
      console.log(`Supabase auth event is ${event}`);
      dispatch(setUserSession(session));
      dispatch(setUser(session?.user ?? null));

      // Call the function to handle user updates
      handleUserUpdates(dispatch, session);
    })
```
It seems after auth state change, any immediate supabase database calls hang after that. In my case, the auth.updateUser() function would not return a response.

Hopefully this helps someone.

**Comment by GaryAustin1** [comment]
I just edited the title of this issue to reflect the root problem so users might spot it quicker.

This bug is causing user major loss of time as it is not documented you can't make calls in onAuthStateChange and the symptom they see is in a completely different part of code.

This should be documented in the docs and fixed or addressed as soon as possible.  It is a multi-hour loss of time for users trying to figure out and/or reaching out for help.

**Comment by brycegoh** [comment]
Big ups to OP we have this problem too m8

**Comment by hf** [comment]
> This should be documented in the docs and fixed or addressed as soon as possible. It is a multi-hour loss of time for users trying to figure out and/or reaching out for help.

@GaryAustin1 Can you point to where you would want this to be present in docs. Trust me, I've tried to address the issue on a technical level -- unfortunately it's not possible because of JavaScript, browsers, NextJS and all the other runtimes.

Alternatively, we can maybe turn off the await handling or make it configurable. Please let me know what works best for you.

**Comment by GaryAustin1** [comment]
@hf It should be documented here https://supabase.com/docs/reference/javascript/auth-onauthstatechange at least to not make supabase calls in that function handler and what happens if you do.   

I added a note in troubleshooting https://github.com/orgs/supabase/discussions/19058 awhile back that I can clean up as it says "until this is fixed".

Not sure the user suggestion above to move calls to a function is the recommended way to avoid this or not: 
```js
supabase.auth.onAuthStateChange(async (event, session) => {
      console.log(`Supabase auth event is ${event}`);
      dispatch(setUserSession(session));
      dispatch(setUser(session?.user ?? null));

      // Call the function to handle user updates
      handleUserUpdates(dispatch, session);
    })
```

**Comment by hf** [comment]
Docs changes here: https://github.com/supabase/supabase/pull/19902

**Comment by justinkchen** [comment]
I removed my onAuthStateChange code and still have getSession hang on supabase-js 2.39.2

**Comment by ScreamZ** [comment]
> Docs changes here: [supabase/supabase#19902](https://github.com/supabase/supabase/pull/19902)

I can confirm this is working see my comment https://github.com/nuxt-modules/supabase/issues/273#issuecomment-2051932773

**Comment by CarrettaRiccardo** [comment]
I found that you have to call every function for auth (signIn, signOut, ...) in client-side code (using a client supabase) in order for it to work properly.
Having things like server actions in the server using a server supabase instance will not trigger an `onAuthStateChange` client side, even tho in that case there are some weird behaviours, like when changing browser tab

**Comment by kelvanb97** [comment]
This is an ongoing issue. Our team fixed this issue by removing all instances of `supabase.auth.onAuthStateChange` other than our `AuthProvider`. In other words we only have implemented one `supabase.auth.onAuthStateChange` call. Our solution is not very DRY but it is reliable.

Without fully debugging the issue there seems to be a lock conflict when there is an auth event emitted to multiple `onAuthStateChange` subscriptions.

**Comment by tothambrus11** [comment]
If it's technically possible, it would be great if the system could detect if a supabase operation has started inside the callback function, and at least log an error to the console, showing the real source of the issue (maybe pointing to this thread or docs).

E.g. 
```js

let globalSupabaseStuffUsed = false;

function onAuthStateChange(callback) {
  ...
  globalSupabaseStuffUsed = false;
  callback(...);
  if(globalSupabaseStuffUsed){
     console.error("Supabase functions were called inside onAuthStateChange. See https://github.com/supabase/auth-js/issues/762");
  }
}

function anySupabaseOperation(...){
   globalSupabaseStuffUsed = true;
   ...
}
```

**Comment by ChronSyn** [comment]
Even though I've not been making supabase-specific calls in the callback, I was still having issues.

I'm still testing to see if the issue is resolved, but specific to my use case, I think my resolution to this was to not try to do any `await` logic in the callback and not marking the callback as `async`.

**Comment by Razikus** [comment]
Hello
```
try {
  const lockState = await navigator.locks.query();
  console.log('Held locks:', lockState.held);
  console.log('Pending locks:', lockState.pending);
  
  // To release locks if needed:
  for (const lock of lockState.held) {
    await navigator.locks.request(lock.name, { steal: true }, () => {});
  }
} catch (error) {
  console.error('Error querying locks:', error);
}
```
This is my workaround to this issue
Supabase is double locking somehow, and the only way is to unlock yourself

**Comment by holys** [comment]
https://supabase.com/docs/reference/javascript/auth-signup

**Comment by kelvanb97** [comment]
> https://supabase.com/docs/reference/javascript/auth-onauthstatechange

Lol, very good 👍 the broken docs 🎉

**Comment by react-ghost** [comment]
Not sure I'm really helping here, but wanted to share that I had a very obscure-to-debug problem in a desktop application I'm building with Next.JS and Tauri, where an `async` function inside the `onAuthStateChanged` callback caused a subsequent supabase call to hang, but the kicker is it was only when the desktop app was built for Windows, so I was none the wiser until I got on a windows machine and developed there for a while and played around with the code.

In any event, I realized there was no reason the function should be asynchronous inside the callback, so removing it solved the problem.

**Comment by MattGyver** [comment]
> @hf It should be documented here https://supabase.com/docs/reference/javascript/auth-onauthstatechange at least to not make supabase calls in that function handler and what happens if you do.
> 
> I added a note in troubleshooting https://github.com/orgs/supabase/discussions/19058 awhile back that I can clean up as it says "until this is fixed".
> 
> Not sure the user suggestion above to move calls to a function is the recommended way to avoid this or not:
> 
> ```js
> supabase.auth.onAuthStateChange(async (event, session) => {
>       console.log(`Supabase auth event is ${event}`);
>       dispatch(setUserSession(session));
>       dispatch(setUser(session?.user ?? null));
> 
>       // Call the function to handle user updates
>       handleUserUpdates(dispatch, session);
>     })
> ```

This issue still seems to be around in the latest versions of supabase. I've spent a couple of days trying to figure out what was causing the await to never return, sure enough it was because I was calling it inside of onAuthStateChange. A lot of hours wasted. :(

**Comment by dan98** [comment]
Having the same issue

**Comment by GaryAustin1** [comment]
@dan98 You have to follow the important note in the docs to avoid this:  https://supabase.com/docs/reference/javascript/auth-onauthstatechange

**Comment by jerdmeyer** [comment]
I just spent a good month trying to figure out what the heck was going on in my app that suddenly stopped allowing user's to interact with Supabase after having the app opened for a while.

Thanks to @GaryAustin1 for helping me out on the Supabase Discord channel. This should really be fixed to not cause major headaches to others moving forward. Or make the docs have a big WARNING: text to make it more clear.

**Comment by hades200082** [comment]
Just spent hours trying to figure out why my supabase calls were silently blackholing my app. Turns out its this bug. Open since 2023 ... please fix it!

**Comment by JVini0166** [comment]
I am having the same issue, Supabase just stop working untuil I reopoen my browser

**Comment by Coinners** [comment]
I also have this issue where await supabase.auth.getSession(); will hang indefinitely without throwing

**Comment by Cominis** [comment]
I have same problem with `supabase.auth.signOut()`
As mentioned [here](https://github.com/supabase/auth-js/issues/762#issuecomment-2781060172)

```javascript
supabase.auth.onAuthStateChange((event, session) => {
  setTimeout(async () => {
    // await on other Supabase function here
    // this runs right after the callback has finished
  }, 0)
})
```

this solved my problem

**Comment by jeffjacobsen** [comment]
I spent hours wondering why things weren't working - Glad I found this fix!

**Comment by rggeorge** [comment]
The solution in my case is very easy: get off of supabase.
This kind of bug is simply not acceptable.

**Comment by marcospgp** [comment]
happy 2 year anniversary!
